homeassistant/components/duwi/__init__.py

```diff
@@ -0,0 +1,248 @@
+"""Support for Duwi Smart devices."""
+
+from typing import Any, NamedTuple
+
+from duwi_smarthome_sdk.base_api import CustomerApi, SharingTokenListener
+from duwi_smarthome_sdk.device_scene_models import CustomerDevice, CustomerScene
+from duwi_smarthome_sdk.manager import Manager, SharingDeviceListener
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import __version__
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.dispatcher import dispatcher_send
+
+from .const import (
+    _LOGGER,
+    ACCESS_TOKEN,
+    ADDRESS,
+    APP_KEY,
+    APP_SECRET,
+    CLIENT_MODEL,
+    CLIENT_VERSION,
+    DOMAIN,
+    DUWI_DISCOVERY_NEW,
+    DUWI_HA_SIGNAL_UPDATE_ENTITY,
+    DUWI_SCENE_UPDATE,
+    HOUSE_KEY,
+    HOUSE_NAME,
+    HOUSE_NO,
+    PASSWORD,
+    PHONE,
+    REFRESH_TOKEN,
+    SUPPORTED_PLATFORMS,
+    WS_ADDRESS,
+)
+
+type DuwiConfigEntry = ConfigEntry[HomeAssistantDuwiData]
+
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(domain=DOMAIN)
+
+
+class HomeAssistantDuwiData(NamedTuple):
+    """Duwi data stored in the Home Assistant data object."""
+
+    manager: Manager
+    listener: SharingDeviceListener
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: DuwiConfigEntry) -> bool:
+    """Set up a config entry for the Duwi Smart Devices integration."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    token_listener = TokenListener(hass, entry)
+    manager: Manager = Manager(
+        customer_api=CustomerApi(
+            address=entry.data[ADDRESS],
+            ws_address=entry.data[WS_ADDRESS],
+            app_key=entry.data[APP_KEY],
+            app_secret=entry.data[APP_SECRET],
+            house_no=entry.data[HOUSE_NO],
+            house_name=entry.data[HOUSE_NAME],
+            access_token=entry.data[ACCESS_TOKEN],
+            refresh_token=entry.data[REFRESH_TOKEN],
+            client_version=CLIENT_VERSION,
+            client_model=CLIENT_MODEL,
+            app_version=__version__,
+        ),
+        house_key=entry.data.get(HOUSE_KEY),
+        token_listener=token_listener,
+    )
+
+    # Fetch devices
+    await manager.init_manager()
+    await manager.update_device_cache()
+    listener = DeviceListener(hass, manager)
+    manager.add_device_listener(listener)
+
+    ids = []
+    if hasattr(entry, "runtime_data") and entry.runtime_data is not None:
+        old_manager = entry.runtime_data.manager
+        ids = await compare_manager(old_manager, manager)
+
+    entry.runtime_data = HomeAssistantDuwiData(manager=manager, listener=listener)
+
+    # Clean up inappropriate devices
+    await cleanup_device_registry(hass, entry, ids)
+    hass.data[DOMAIN].setdefault("existing_house", []).append(entry.data[HOUSE_NO])
```

Comment:
If the house number has already been added, it will not be displayed in the selector house section of confic_flow
This is to prevent users from adding houses repeatedly (we have always recommended that users configure only one house in the HA system)

Reply:
Okay but we have other ways to do this than by storing this in hass.data. rather just iterate over the config entries in the config flow

Reply:
Thank you for the reminder. I was not aware of this solution before, but I will make modifications in the future

Reply:
Thank you very much for the solution you provided

I have considered it here because my previous idea was to make this' existing'house 'a globally visible variable, in order to let other instances know which houses have been added. I tested that entry.data is globally invisible

I also tried removing this existing'house afterwards and found that devices with the same device type and device ID will not be added to the system again. Therefore, I directly deleted this existing'house variable without storing it

Reply:
[#08198dc](https://github.com/home-assistant/core/pull/125803/commits/08198dcb3671940cff044284f405fceb2bd5e3ae)